File: sdks/typescript/src/commands.ts
export interface ProcessResult {
  exitCode: number;
  stdout: string;
  stderr: string;
}

export interface RunOpts {
  timeout?: number;
  env?: Record<string, string>;
  cwd?: string;
  onStdout?: (data: string) => void;
  onStderr?: (data: string) => void;
  /** Allocate a PTY for real-time unbuffered output (npm, apt, pip, etc.). */
  tty?: boolean;
  /** Run the command in the background. Returns a CommandHandle immediately
   *  instead of waiting for the process to exit. Use for long-running servers,
   *  dev tools, etc. The process keeps running even if you don't hold a reference. */
  background?: boolean;
}

export interface ExecChunk {
  stream: "stdout" | "stderr";
  data: string;
}

/** Handle for a background process. Allows waiting, killing, or disconnecting. */
export class CommandHandle {
  readonly sessionId: string;
  private _ws: WebSocket | null = null;
  private _killed = false;

  constructor(
    sessionId: string,
    private apiUrl: string,
    private headers: Record<string, string>,
    private sandboxId: string,
    ws: WebSocket | null,
    private onStdout?: (data: string) => void,
    private onStderr?: (data: string) => void,
  ) {
    this.sessionId = sessionId;
    this._ws = ws;
  }

  /** Send data to the process's stdin. */
  sendInput(data: string): void {
    if (this._ws && this._ws.readyState === this._ws.OPEN) {
      this._ws.send(data);
    }
  }

  /** Kill the background process. */
  async kill(): Promise<void> {
    if (this._killed) return;
    this._killed = true;
    this._ws?.close();
    await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/pty/${this.sessionId}`, {
      method: "DELETE",
      headers: this.headers,
    }).catch(() => {});
  }

  /** Disconnect from the process without killing it. The process keeps running
   *  in the sandbox. You can reconnect later via `commands.connect()`. */
  disconnect(): void {
    this._ws?.close();
    this._ws = null;
  }

  /** Wait for the background process to exit. Resolves when the WebSocket closes. */
  async wait(): Promise<void> {
    if (!this._ws) return;
    const ws = this._ws;
    if (ws.readyState === ws.CLOSED) return;
    await new Promise<void>((resolve) => {
      ws.addEventListener("close", () => resolve(), { once: true });
    });
  }
}

export class Commands {
  constructor(
    private apiUrl: string,
    private apiKey: string,
    private sandboxId: string,
    private token: string = "",
  ) {}

  private get headers(): Record<string, string> {
    const h: Record<string, string> = { "Content-Type": "application/json" };
    if (this.token) {
      h["Authorization"] = `Bearer ${this.token}`;
    } else if (this.apiKey) {
      h["X-API-Key"] = this.apiKey;
    }
    return h;
  }

  /**
   * Run a command in the sandbox.
   *
   * - Default: waits for completion, returns ProcessResult.
   * - `{ background: true }`: starts the process and returns a CommandHandle immediately.
   *   The process runs in a PTY session and survives client disconnect.
   * - `onStdout`/`onStderr`: stream output in real-time via SSE (foreground only).
   */
  run(command: string, opts: RunOpts & { background: true }): Promise<CommandHandle>;
  run(command: string, opts?: RunOpts & { background?: false }): Promise<ProcessResult>;
  run(command: string, opts: RunOpts = {}): Promise<ProcessResult | CommandHandle> {
    if (opts.background) {
      return this._runBackground(command, opts);
    }
    if (opts.onStdout || opts.onStderr || opts.tty) {
      return this._runStreaming(command, opts);
    }
    return this._runSimple(command, opts);
  }

  /**
   * Connect to an already-running background process by its session ID.
   */
  async connect(sessionId: string, opts?: {
    onStdout?: (data: string) => void;
    onStderr?: (data: string) => void;
  }): Promise<CommandHandle> {
    const ws = this._connectWebSocket(sessionId, opts?.onStdout, opts?.onStderr);
    return new CommandHandle(sessionId, this.apiUrl, this.headers, this.sandboxId, ws, opts?.onStdout, opts?.onStderr);
  }

  /** List active PTY sessions (background processes). */
  // Note: This relies on the existing PTY list if/when the backend exposes it.
  // For now, callers track session IDs themselves.

  /** Kill a background process by session ID. */
  async kill(sessionId: string): Promise<void> {
    await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/pty/${sessionId}`, {
      method: "DELETE",
      headers: this.headers,
    }).catch(() => {});
  }

  // --- Private implementations ---

  private async _runSimple(command: string, opts: RunOpts): Promise<ProcessResult> {
    const timeout = opts.timeout ?? 60;

    const body: Record<string, unknown> = {
      cmd: command,
      timeout,
    };
    if (opts.env) body.envs = opts.env;
    if (opts.cwd) body.cwd = opts.cwd;

    const controller = new AbortController();
    const timeoutId = globalThis.setTimeout(() => controller.abort(), (timeout + 5) * 1000);

    try {
      const resp = await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/commands`, {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      if (!resp.ok) {
        const text = await resp.text();
        throw new Error(`Command failed: ${resp.status} ${text}`);
      }

      return await resp.json();
    } finally {
      globalThis.clearTimeout(timeoutId);
    }
  }

  private async _runStreaming(command: string, opts: RunOpts): Promise<ProcessResult> {
    const timeout = opts.timeout ?? 60;

    const body: Record<string, unknown> = {
      cmd: command,
      timeout,
    };
    if (opts.env) body.envs = opts.env;
    if (opts.cwd) body.cwd = opts.cwd;
    if (opts.tty) body.tty = true;

    const controller = new AbortController();
    const idleMs = (timeout + 30) * 1000;
    let idleTimer = globalThis.setTimeout(() => controller.abort(), idleMs);
    const resetIdle = () => {
      globalThis.clearTimeout(idleTimer);
      idleTimer = globalThis.setTimeout(() => controller.abort(), idleMs);
    };

    let stdout = "";
    let stderr = "";
    let exitCode = 0;

    try {
      const resp = await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/exec`, {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      if (!resp.ok) {
        const text = await resp.text();
        throw new Error(`Command failed: ${resp.status} ${text}`);
      }

      if (!resp.body) {
        throw new Error("No response body for streaming");
      }

      const reader = resp.body.getReader();
      const decoder = new TextDecoder();
      let buffer = "";

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        resetIdle();

        buffer += decoder.decode(value, { stream: true });
        const parts = buffer.split("\n\n");
        buffer = parts.pop() || "";

        for (const part of parts) {
          const event = parseSSEEvent(part);
          if (!event) continue;

          if (event.type === "stdout" || event.type === "stderr") {
            const data = JSON.parse(event.data);
            if (event.type === "stdout") {
              stdout += data.data;
              opts.onStdout?.(data.data);
            } else {
              stderr += data.data;
              opts.onStderr?.(data.data);
            }
          } else if (event.type === "exit") {
            const data = JSON.parse(event.data);
            exitCode = data.exit_code;
          } else if (event.type === "error") {
            const data = JSON.parse(event.data);
            throw new Error(data.error);
          }
        }
      }
    } finally {
      globalThis.clearTimeout(idleTimer);
    }

    return { exitCode, stdout, stderr };
  }

  private async _runBackground(command: string, opts: RunOpts): Promise<CommandHandle> {
    // Create a PTY session
    const resp = await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/pty`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify({
        cols: 120,
        rows: 40,
      }),
    });

    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Failed to start background process: ${resp.status} ${text}`);
    }

    const data = await resp.json();
    const sessionId: string = data.sessionID;

    // Connect WebSocket for output
    const ws = this._connectWebSocket(sessionId, opts.onStdout, opts.onStderr);

    // Wait for WebSocket to open, then send the command
    await new Promise<void>((resolve, reject) => {
      ws.addEventListener("open", () => resolve(), { once: true });
      ws.addEventListener("error", (e) => reject(new Error("WebSocket connection failed")), { once: true });
    });

    // Send command + newline to execute it
    const envPrefix = opts.env
      ? Object.entries(opts.env).map(([k, v]) => `export ${k}=${shellEscape(v)}`).join(" && ") + " && "
      : "";
    const cdPrefix = opts.cwd ? `cd ${shellEscape(opts.cwd)} && ` : "";
    ws.send(`${envPrefix}${cdPrefix}${command}\n`);

    return new CommandHandle(sessionId, this.apiUrl, this.headers, this.sandboxId, ws, opts.onStdout, opts.onStderr);
  }

  private _connectWebSocket(
    sessionId: string,
    onStdout?: (data: string) => void,
    onStderr?: (data: string) => void,
  ): WebSocket {
    const wsUrl = this.apiUrl
      .replace("http://", "ws://")
      .replace("https://", "wss://");
    const ws = new WebSocket(`${wsUrl}/sandboxes/${this.sandboxId}/pty/${sessionId}`);
    ws.binaryType = "arraybuffer";

    if (onStdout) {
      const decoder = new TextDecoder();
      ws.onmessage = (event) => {
        const text = event.data instanceof ArrayBuffer
          ? decoder.decode(new Uint8Array(event.data))
          : event.data as string;
        // PTY merges stdout/stderr, so everything comes as stdout
        onStdout(text);
      };
    }

    return ws;
  }
}

function parseSSEEvent(raw: string): { type: string; data: string } | null {
  let type = "message";
  let data = "";

  for (const line of raw.split("\n")) {
    if (line.startsWith("event: ")) {
      type = line.slice(7).trim();
    } else if (line.startsWith("data: ")) {
      data = line.slice(6);
    }
  }

  if (!data) return null;
  return { type, data };
}

function shellEscape(s: string): string {
  return "'" + s.replace(/'/g, "'\\''") + "'";
}

File: sdks/python/opencomputer/commands.py
"""Command execution inside a sandbox."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Callable

import httpx


@dataclass
class ProcessResult:
    """Result of a command execution."""

    exit_code: int
    stdout: str
    stderr: str


@dataclass
class ExecChunk:
    """A single chunk of streaming command output."""

    stream: str  # "stdout" or "stderr"
    data: str


@dataclass
class CommandHandle:
    """Handle for a background process running in the sandbox."""

    session_id: str
    _client: httpx.AsyncClient = field(repr=False)
    _sandbox_id: str = field(repr=False)

    async def kill(self) -> None:
        """Kill the background process."""
        await self._client.delete(
            f"/sandboxes/{self._sandbox_id}/pty/{self.session_id}",
        )

    async def send_input(self, data: str) -> None:
        """Send data to the process's stdin (requires websocket — placeholder)."""
        raise NotImplementedError(
            "Use the PTY websocket to send stdin to background processes"
        )


@dataclass
class Commands:
    """Command execution for a sandbox."""

    _client: httpx.AsyncClient
    _sandbox_id: str

    async def run(
        self,
        command: str,
        timeout: int = 60,
        env: dict[str, str] | None = None,
        cwd: str | None = None,
        on_stdout: Callable[[str], None] | None = None,
        on_stderr: Callable[[str], None] | None = None,
        tty: bool = False,
        background: bool = False,
    ) -> ProcessResult | CommandHandle:
        """Run a command in the sandbox.

        Args:
            command: The command to run.
            timeout: Timeout in seconds (default 60).
            env: Optional environment variables.
            cwd: Optional working directory.
            on_stdout: Callback for each stdout chunk (enables streaming).
            on_stderr: Callback for each stderr chunk (enables streaming).
            tty: Allocate a PTY for real-time unbuffered output.
            background: Start the process in the background and return a
                CommandHandle immediately. The process runs in a PTY session
                and survives client disconnect.

        Returns:
            ProcessResult for foreground commands, CommandHandle for background.
        """
        if background:
            return await self._run_background(command, env=env, cwd=cwd)

        if on_stdout or on_stderr or tty:
            return await self._run_streaming(
                command, timeout=timeout, env=env, cwd=cwd,
                on_stdout=on_stdout, on_stderr=on_stderr, tty=tty,
            )

        return await self._run_simple(command, timeout=timeout, env=env, cwd=cwd)

    async def connect(self, session_id: str) -> CommandHandle:
        """Connect to an already-running background process by session ID."""
        return CommandHandle(
            session_id=session_id,
            _client=self._client,
            _sandbox_id=self._sandbox_id,
        )

    async def kill(self, session_id: str) -> None:
        """Kill a background process by session ID."""
        await self._client.delete(
            f"/sandboxes/{self._sandbox_id}/pty/{session_id}",
        )

    # --- Private implementations ---

    async def _run_simple(
        self,
        command: str,
        timeout: int = 60,
        env: dict[str, str] | None = None,
        cwd: str | None = None,
    ) -> ProcessResult:
        body: dict[str, Any] = {
            "cmd": command,
            "timeout": timeout,
        }
        if env:
            body["envs"] = env
        if cwd:
            body["cwd"] = cwd

        resp = await self._client.post(
            f"/sandboxes/{self._sandbox_id}/commands",
            json=body,
            timeout=timeout + 5,
        )
        resp.raise_for_status()
        data = resp.json()

        return ProcessResult(
            exit_code=data.get("exitCode", -1),
            stdout=data.get("stdout", ""),
            stderr=data.get("stderr", ""),
        )

    async def _run_streaming(
        self,
        command: str,
        timeout: int = 60,
        env: dict[str, str] | None = None,
        cwd: str | None = None,
        on_stdout: Callable[[str], None] | None = None,
        on_stderr: Callable[[str], None] | None = None,
        tty: bool = False,
    ) -> ProcessResult:
        body: dict[str, Any] = {
            "cmd": command,
            "timeout": timeout,
        }
        if env:
            body["envs"] = env
        if cwd:
            body["cwd"] = cwd
        if tty:
            body["tty"] = True

        stdout_parts: list[str] = []
        stderr_parts: list[str] = []
        exit_code = -1

        read_timeout = timeout + 30
        async with self._client.stream(
            "POST",
            f"/sandboxes/{self._sandbox_id}/exec",
            json=body,
            timeout=httpx.Timeout(5.0, read=read_timeout),
        ) as resp:
            resp.raise_for_status()
            buffer = ""
            async for raw in resp.aiter_text():
                buffer += raw
                while "\n\n" in buffer:
                    event_str, buffer = buffer.split("\n\n", 1)
                    event = _parse_sse_event(event_str)
                    if event is None:
                        continue

                    if event["type"] in ("stdout", "stderr"):
                        payload = json.loads(event["data"])
                        chunk_data = payload["data"]
                        if event["type"] == "stdout":
                            stdout_parts.append(chunk_data)
                            if on_stdout:
                                on_stdout(chunk_data)
                        else:
                            stderr_parts.append(chunk_data)
                            if on_stderr:
                                on_stderr(chunk_data)
                    elif event["type"] == "exit":
                        payload = json.loads(event["data"])
                        exit_code = payload["exit_code"]
                    elif event["type"] == "error":
                        payload = json.loads(event["data"])
                        raise RuntimeError(f"Exec error: {payload['error']}")

        return ProcessResult(
            exit_code=exit_code,
            stdout="".join(stdout_parts),
            stderr="".join(stderr_parts),
        )

    async def _run_background(
        self,
        command: str,
        env: dict[str, str] | None = None,
        cwd: str | None = None,
    ) -> CommandHandle:
        # Create a PTY session
        resp = await self._client.post(
            f"/sandboxes/{self._sandbox_id}/pty",
            json={"cols": 120, "rows": 40},
        )
        resp.raise_for_status()
        session_id = resp.json()["sessionID"]

        # Build the command with env/cwd prefixes
        parts = []
        if env:
            for k, v in env.items():
                parts.append(f"export {k}={_shell_escape(v)}")
        if cwd:
            parts.append(f"cd {_shell_escape(cwd)}")
        parts.append(command)
        full_cmd = " && ".join(parts)

        # Send command via a short-lived websocket
        import websockets  # type: ignore

        ws_url = str(self._client.base_url).replace("http://", "ws://").replace("https://", "wss://")
        ws_endpoint = f"{ws_url}/sandboxes/{self._sandbox_id}/pty/{session_id}"

        async with websockets.connect(ws_endpoint) as ws:
            await ws.send(full_cmd + "\n")

        return CommandHandle(
            session_id=session_id,
            _client=self._client,
            _sandbox_id=self._sandbox_id,
        )


def _parse_sse_event(raw: str) -> dict[str, str] | None:
    """Parse a single SSE event block into type + data."""
    event_type = "message"
    data = ""
    for line in raw.split("\n"):
        if line.startswith("event: "):
            event_type = line[7:].strip()
        elif line.startswith("data: "):
            data = line[6:]
    if not data:
        return None
    return {"type": event_type, "data": data}


def _shell_escape(s: str) -> str:
    return "'" + s.replace("'", "'\\''") + "'"
